`associated_types.hpp` compilation failure when generic `operator-` is visible (gcc only)

With the `0.9.1` release and today's `HEAD` (67d5645) I have started to receive an error when compiling with g++.

```
#include <range/v3/all.hpp>

template <class A, class B>
constexpr auto operator-(A a, B) {
  return a;
}

int main(int, char * const[]) {
  std::vector<int> data;
  ranges::views::all(data) | ranges::actions::sort;
  return 0;
}
```

This fails with `g++ 9.2`.

```
ldalessa@portland:~/cswarm/mrwt/debug$ g++ --version
g++ (Debian 9.2.1-8) 9.2.1 20190909
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

ldalessa@portland:~/cswarm/mrwt/debug$ g++ -I/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include -std=gnu++17 -c -o temp.o temp.cpp 
In file included from /home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/iterator/access.hpp:22,
                 from /home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/iterator/concepts.hpp:30,
                 from /home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/range/concepts.hpp:28,
                 from /home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/action/concepts.hpp:23,
                 from /home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/action/action.hpp:23,
                 from /home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/action.hpp:17,
                 from /home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/all.hpp:17,
                 from temp.cpp:1:
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/std/detail/associated_types.hpp: In substitution of ‘template<class T, T <anonymous> > constexpr bool ranges::detail::is_integral_(long int) [with T = ranges::ref_view<std::vector<int> >; T <anonymous> = <missing>]’:
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/std/detail/associated_types.hpp:132:67:   required by substitution of ‘template<class T> struct ranges::detail::incrementable_traits_2_<T, ranges::detail::always_<void, int [is_integral_<ranges::detail::difference_result_t<T> >(0)]> > [with T = ranges::ref_view<std::vector<int> >]’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/std/detail/associated_types.hpp:139:16:   required from ‘struct ranges::detail::incrementable_traits_1_<ranges::ref_view<std::vector<int> >, void>’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/std/detail/associated_types.hpp:162:12:   required from ‘struct ranges::incrementable_traits<ranges::ref_view<std::vector<int> > >’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/iterator/concepts.hpp:190:5:   required by substitution of ‘template<class C_> static constexpr decltype (((& C_::Requires_<ref_view<std::vector<int, std::allocator<int> > > >), true)) ranges::weakly_incrementable_concept::Eval<ranges::ref_view<std::vector<int> > >::impl<C_>(int) [with C_ = ranges::weakly_incrementable_concept]’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/iterator/concepts.hpp:190:5:   required from ‘constexpr ranges::weakly_incrementable_concept::Eval<I>::operator bool() const [with I = ranges::ref_view<std::vector<int> >]’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/iterator/concepts.hpp:190:5:   [ skipping 19 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/functional/concepts.hpp:27:5:   required by substitution of ‘template<class C_> static constexpr decltype (((& C_::Requires_<sort_fn&, ref_view<std::vector<int, std::allocator<int> > > >), true)) ranges::invocable_concept::Eval<ranges::actions::sort_fn&, ranges::ref_view<std::vector<int, std::allocator<int> > > >::impl<C_>(int) [with C_ = ranges::invocable_concept]’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/functional/concepts.hpp:27:5:   required from ‘constexpr ranges::invocable_concept::Eval<Fun, Args>::operator bool() const [with Fun = ranges::actions::sort_fn&; Args = {ranges::ref_view<std::vector<int, std::allocator<int> > >}]’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/functional/concepts.hpp:27:5:   required from ‘constexpr const bool ranges::invocable<ranges::actions::sort_fn&, ranges::ref_view<std::vector<int, std::allocator<int> > > >’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/action/action.hpp:75:20:   required by substitution of ‘template<class Rng, class Act> static concepts::return_t<decltype (ranges::invoke(declval<ranges::actions::sort_fn&>(), declval<Rng>())), typename std::enable_if<(((range<Rng> && invocable<ranges::actions::sort_fn&, Rng>) && (! std::is_reference<_Tp>::value)) && concepts::detail::CPP_true(concepts::detail::Nil{})), void>::type> ranges::actions::action<ranges::actions::sort_fn>::pipe<Rng, Act>(Rng&&, Act&&) [with Rng = ranges::ref_view<std::vector<int> >; Act = ranges::actions::action<ranges::actions::sort_fn>&]’
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/range/v3/functional/pipeable.hpp:63:53:   required from ‘constexpr auto ranges::operator|(Arg&&, Pipe) [with Arg = ranges::ref_view<std::vector<int> >; Pipe = ranges::actions::action<ranges::actions::sort_fn>; CPP_true_ = std::integral_constant<bool, true>; typename std::enable_if<(((! is_pipeable_v<Arg>) && is_pipeable_v<Pipe>) && CPP_true_{}), int>::type <anonymous> = 0]’
temp.cpp:13:47:   required from here
/home/ldalessa/cswarm/mrwt/debug/_deps/rangev3-src/include/std/detail/associated_types.hpp:97:34: error: non-type template parameters of class type only available with ‘-std=c++2a’ or ‘-std=gnu++2a’
   97 |         template<typename T, T = 1>
      |                                  ^
```

It's fine if I add `-fconcepts`, and it's fine with clang 8.0.1.

--- a/include/range/v3/detail/config.hpp
+++ b/include/range/v3/detail/config.hpp
@@ -331,10 +331,16 @@ namespace ranges
 #if __GNUC__ == 7 || __GNUC__ == 8
 #define RANGES_WORKAROUND_GCC_91525 /* Workaround strange GCC ICE */
 #endif
-#if __GNUC__ >= 9 && defined(__cpp_concepts)
-#define RANGES_WORKAROUND_GCC_89953
+#if __GNUC__ >= 9
+#ifdef __cpp_concepts
+#define RANGES_WORKAROUND_GCC_89953 // ICE in nothrow_spec_p, at cp/except.c:1244
+#endif
+#if __cplusplus <= RANGES_CXX_STD_17
+#define RANGES_WORKAROUND_GCC_91923 // Failure-to-SFINAE with class type NTTP in C++17
 #endif
 #endif
+#endif
+
 
 #else
 #define RANGES_DIAGNOSTIC_PUSH

--- a/include/range/v3/detail/config.hpp
+++ b/include/range/v3/detail/config.hpp
@@ -335,7 +335,7 @@ namespace ranges
 #ifdef __cpp_concepts
 #define RANGES_WORKAROUND_GCC_89953 // ICE in nothrow_spec_p, at cp/except.c:1244
 #endif
-#if __cplusplus <= RANGES_CXX_STD_17
+#if __GNUC__ == 9 && __GNUC_MINOR__ < 3 && __cplusplus <= RANGES_CXX_STD_17
 #define RANGES_WORKAROUND_GCC_91923 // Failure-to-SFINAE with class type NTTP in C++17
 #endif
 #endif

--- a/include/std/detail/associated_types.hpp
+++ b/include/std/detail/associated_types.hpp
@@ -126,7 +126,8 @@ namespace ranges
         template<typename T>
         struct incrementable_traits_2_<
             T,
-#if defined(_MSC_VER) && !defined(__clang__) && !defined(__EDG__)
+#if (defined(_MSC_VER) && !defined(__clang__) && !defined(__EDG__)) || \
+    defined(RANGES_WORKAROUND_GCC_91923)
             std::enable_if_t<std::is_integral_v<difference_result_t<T>>>>
 #else  // ^^^ MSVC / not MSVC vvv
             always_<void, int[is_integral_<difference_result_t<T>>(0)]>>

--- a/test/bug1335.cpp
+++ b/test/bug1335.cpp
@@ -1,12 +1,14 @@
 #include <vector>
 #include <range/v3/action/sort.hpp>
 
-template <class A, class B>
-constexpr auto operator-(A a, B) {
-  return a;
+template<typename A, typename B>
+constexpr auto operator-(A a, B)
+{
+    return a;
 }
 
-int main() {
-  std::vector<int> data;
-  data |= ranges::actions::sort;
+int main()
+{
+    std::vector<int> data;
+    data |= ranges::actions::sort;
 }

--- a/test/bug1335.cpp
+++ b/test/bug1335.cpp
@@ -1,0 +1,12 @@
+#include <vector>
+#include <range/v3/action/sort.hpp>
+
+template <class A, class B>
+constexpr auto operator-(A a, B) {
+  return a;
+}
+
+int main() {
+  std::vector<int> data;
+  data |= ranges::actions::sort;
+}

